Modules/_ssl.c

```diff
@@ -4600,6 +4749,45 @@ _ssl__SSLContext_set_default_verify_paths_impl(PySSLContext *self)
     Py_RETURN_NONE;
 }
 
+/*[clinic input]
+_ssl._SSLContext.set_ech_config
+
+    ech_config: Py_buffer
+    /
+
+Set the ECH configuration on the SSL context.
+
+The echconfig parameter should be a bytes-like object containing the raw ECH configuration.
+[clinic start generated code]*/
+
+static PyObject *
+_ssl__SSLContext_set_ech_config_impl(PySSLContext *self,
+                                     Py_buffer *ech_config)
+/*[clinic end generated code: output=cea53188b338cf80 input=09f9a38c5aaef091]*/
+{
+#ifdef OPENSSL_ECH
+    OSSL_ECHSTORE *es = NULL;
+    BIO *es_in = NULL;
+
+    if ((es_in = BIO_new_mem_buf(ech_config->buf, ech_config->len)) == NULL
+        || (es = OSSL_ECHSTORE_new(NULL, NULL)) == NULL
+        || OSSL_ECHSTORE_read_echconfiglist(es, es_in) != 1
+        || SSL_CTX_set1_echstore(self->ctx, es) != 1) {
+        _setSSLError(get_state_ctx(self), NULL, 0, __FILE__, __LINE__);
```

Comment:
Break down this big if into multiple ones and use a goto error with an error label doing the cleanup.